jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/lock/RuntimeManagerLockWatcher.java

```diff
@@ -0,0 +1,147 @@
+
+    private Long watchLockPolling;
+    private Long maxLockProcessingTime;
+
+    private ExecutorService executorService;
+    private Queue<LockWatchInfo> locksWatched;
+
+    class RuntimeManagerLockReaper implements Callable<Void> {
+
+        @Override
+        public Void call() throws Exception {
+            while(!Thread.currentThread().isInterrupted()) {
+                long maxTimeToWait = reap();
+                Thread.sleep(maxTimeToWait);
+            }
+            return null;
+        }
+
+    }
+
+    public long reap() {
+        synchronized (locksWatched) {
+            while(!locksWatched.isEmpty()) {
+                LockWatchInfo info = locksWatched.peek();
+                long idleTime = info.getCurrentProcessingTime();
+                if(maxLockProcessingTime > idleTime) {
+                    return maxLockProcessingTime - idleTime;
+                }
+                try {
+                    long timeExceeded = idleTime - maxLockProcessingTime;
+                    logger.info("Max process time lock exceeded for process instance id {} by {} ms. Trying forceful release lock {}", info.processInstanceId, timeExceeded, info.lock);
+                    info.lock.forceUnlock();
+                } catch(UnsupportedOperationException e) {
+                    logger.warn("Runtime manager lock implementation {} does not allow self release for process instance id {}", info.lock.getClass().getName(), info.processInstanceId);
+                }
+                // stop watching this lock. We gave already the information
+                locksWatched.poll();
+            }
+        }
+
+        return watchLockPolling;
+    }
+
+    
+    public RuntimeManagerLockWatcher() {
+        this(Long.getLong("org.kie.jbpm.lock.polling", 0L), Long.getLong("org.kie.jbpm.lock.maxProcessingTime", 60000L));
+    }
+    
+    public RuntimeManagerLockWatcher(long watchLockPolling, long maxLockProcessingTime) {
+        this.watchLockPolling = watchLockPolling;
+        this.maxLockProcessingTime = maxLockProcessingTime;
+
+        locksWatched = new PriorityQueue<>();
+        // if no watch polling is set then we don't execute anything
+        if(this.watchLockPolling <= 0) {
+            return;
+        }
+        executorService = Executors.newSingleThreadExecutor();
+        executorService.submit(new RuntimeManagerLockReaper());
+    }
+
+    public void watch(long processInstanceId, RuntimeManagerLock lock) {
+        synchronized (locksWatched) {
+            locksWatched.add(new LockWatchInfo(processInstanceId, lock));
+        }
+
+    }
+
+    public void unwatch(long processInstanceId) {
+        synchronized (locksWatched) {
+            locksWatched.removeIf(e -> e.getProcessInstanceId() == processInstanceId);
+        }
+
+    }
+
+    public void close() {
+        if(watchLockPolling <= 0 || executorService == null) {
+            return;
+        }
+        executorService.shutdown();
```

Comment:
I think this won't send interrupt signal to the already running tasks on the executor. And given the reaper is an infinite loop, it will never finish normally, so we need to use `shutdownNow()` I think as only this cancels already running tasks.

Reply:
you are right